// jser-stat/lib/models/JSerItem.js
// LICENSE : MIT
"use strict";
exports.__esModule = true;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _assert = require("assert");

var _assert2 = _interopRequireDefault(_assert);

var _JSerItemRelatedLink = require("./JSerItemRelatedLink");

var _JSerItemRelatedLink2 = _interopRequireDefault(_JSerItemRelatedLink);

var JSerItem = (function () {
    function JSerItem(item) {
        _classCallCheck(this, JSerItem);

        /** @type {string} */
        this.title = item["title"];
        /** @type {string} */
        this.url = item["url"];
        /** @type {string} */
        this.content = item["content"];
        /** @type {string[]} */
        this.tags = item["tags"] || [];
        /** @type {Date} */
        this.date = new Date(item["date"]);
        var relatedLinks = item["relatedLinks"] || [];
        /** @type {JSerItemRelatedLink[]} */
        this.relatedLinks = relatedLinks.map(function (link) {
            return new _JSerItemRelatedLink2["default"](link);
        });
    }

    /**
     * @param {JSerItem} item
     * @returns {boolean}
     */

    JSerItem.prototype.isEqualItem = function isEqualItem(item) {
        (0, _assert2["default"])(item != null, "item should not be null");
        return this.url === item.url;
    };

    return JSerItem;
})();

exports["default"] = JSerItem;
module.exports = exports["default"];
