// jser-stat/lib/jser-stat.js
// LICENSE : MIT
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _modelsJSerItem = require("./models/JSerItem");

var _modelsJSerItem2 = _interopRequireDefault(_modelsJSerItem);

var _modelsJSerPost = require("./models/JSerPost");

var _modelsJSerPost2 = _interopRequireDefault(_modelsJSerPost);

var _modelsJSerWeek = require("./models/JSerWeek");

var _modelsJSerWeek2 = _interopRequireDefault(_modelsJSerWeek);

var _algoAlgoItem = require("./algo/AlgoItem");

var _algoAlgoItem2 = _interopRequireDefault(_algoAlgoItem);

function filterJSerCategory(article) {
    return /jser/i.test(article.category);
}

var JSerStat = (function () {
    function JSerStat() {
        _classCallCheck(this, JSerStat);

        this._rawItems = require("./data/items.json");
        this._rawPosts = require("./data/posts.json");
        this.items = this._rawItems.map(function (item) {
            return new _modelsJSerItem2["default"](item);
        });
        // JSer カテゴリだけにする
        this.posts = this._rawPosts.filter(filterJSerCategory).map(function (post) {
            return new _modelsJSerPost2["default"](post);
        });
        this.algoItem = new _algoAlgoItem2["default"](this.items);
    }

    _createClass(JSerStat, [{
        key: "getJSerWeeks",
        value: function getJSerWeeks() {
            var _this = this;

            var results = [];
            this.posts.reduce(function (currentPost, nextPost) {
                var jserWeek = new _modelsJSerWeek2["default"](currentPost, nextPost, _this.algoItem);
                results.push(jserWeek);
                return nextPost;
            });
            return results;
        }
    }]);

    return JSerStat;
})();

exports["default"] = JSerStat;
module.exports = exports["default"];
